Nautilus/Assets/ModPrefabCache.cs

```diff
@@ -1,10 +1,74 @@
+using System.Collections.Generic;
+using UnityEngine;
+
 namespace Nautilus.Assets;
 
 /// <summary>
-/// Class used by <see cref="PrefabInfo"/> to store game objects that used as prefabs.
-/// Game objects in cache are inactive and will not be on scene.
+/// Class used by the prefab system to store GameObjects.
+/// Objects in the cache are inactive because they are placed within an inactive parent object.
 /// </summary>
 public static class ModPrefabCache
 {
-    
+    private static ModPrefabCacheInstance _cacheInstance;
+
+    internal record struct Entry(string ClassId, GameObject Prefab);
+
+    /// <summary> Adds the given prefab to the cache. </summary>
+    /// <param name="prefab"> The prefab object that is disabled and cached. </param>
+    public static void AddPrefab(GameObject prefab)
+    {
+        EnsureCacheExists();
+    }
+
+    /// <summary>
+    /// Returns
+    /// </summary>
+    /// <param name="classId"></param>
+    /// <returns></returns>
+    public static bool IsPrefabCached(string classId)
+    {
+        if (_cacheInstance == null)
+            return false;
+        return _cacheInstance.Entries.ContainsKey(classId);
+    }
+
+    /// <summary>
+    /// Returns true if the prefab matching the given <paramref name="classId"/> is found in the cache, where it then assigns the <paramref name="prefab"/> out parameter.
+    /// </summary>
+    /// <param name="classId"></param>
+    /// <param name="prefab"></param>
+    /// <returns></returns>
+    public static bool TryGetPrefabFromCache(string classId, out GameObject prefab)
+    {
+        EnsureCacheExists();
+
+        if (_cacheInstance.Entries.TryGetValue(classId, out var found))
+        {
+            prefab = found.Prefab;
+            return prefab != null;
+        }
+
+        prefab = null;
+        return false;
+    }
+
+    private static void EnsureCacheExists()
+    {
+        if (_cacheInstance != null)
+            return;
+        _cacheInstance = new GameObject("Nautilus.PrefabCache").AddComponent<ModPrefabCacheInstance>();
```

Comment:
Make sure that this gameobject isn't likely to be removed. I can definitely see it needing something like a SceneCleanerPreserve on it to make sure it stays as long as possible.

Reply:
Wouldn't we want the prefab cache to be cleared when exiting to the main menu? I intentionally removed that part.

Reply:
Why would we? It’s just a cache, and I don’t see custom prefabs being randomly removed from the game just based on which save is loaded.

Reply:
I just don't really see a reason for stuff to exist in the background always. Scene changes can also break references etc (although yes preferably you should not be assigning references in your GetGameObject method).

Reply:
Because it’s a cache. If things are cached then they’ll be quicker to get the next time they’re used. Hell, it might even save on load times for the second save because any modded objects or items in the second save may very well be quicker to get because they’ve already been cached

Reply:
What about mods that attempt to unload assets (something we tried with the garg) when unused?

Reply:
Besides the edge cases you have me convinced.

Reply:
Then those are niche use cases, and can still be handled with a RemoveFromCache method (which I believe already exists?)

Reply:
Alright... fair enough